lib/web_catalog/api_extractor.es6.js

```diff
@@ -140,511 +242,440 @@ foam.CLASS({
+      }
+
+      return this.getCtorNamesFromProto_(protoId);
+    },
+    function getCtorNamesFromProto_(protoId) {
+      const og = this.objectGraph;
+      let names = this.classNamesFromGraphPaths ? og.getKeys(protoId)
+            .filter(key => key.endsWith('.prototype'))
+            .map(key => {
+              const parts = key.split('.');
+              return parts[parts.length - 2];
+            }) : [];
+
+      if (this.classNamesFromToString) {
+        let nameFromToString = this.getCtorNameFromToString_(protoId);
+        if (nameFromToString && nameFromToString !== 'Object' &&
+            !names.includes(nameFromToString)) {
+          names.push(nameFromToString);
+        }
+      }
+      return names;
+    },
+    function getCtorNameFromToString_(protoId) {
+      const og = this.objectGraph;
+      let name = og.getToString(protoId);
+      if (!name) return '';
+      const match = name.match(/\[object ([A-Za-z_$][0-9A-Za-z_$]*)\]/);
+      if (match === null) return '';
+      name = match[1];
+      if (name.endsWith('Prototype'))
+        return name.substr(0, name.length - 'Prototype'.length);
+      if (name.endsWith('Constructor'))
+        return name.substr(0, name.length - 'Constructor'.length);
+      return name;
+    },
+    function isFunctionLike_(id) {
+      const og = this.objectGraph;
+      return (!og.isType(id)) &&
+          og.getObjectKeys(id).includes('prototype');
+    },
+    function storeAPIsFromCtorOrLib_(ctorId) {
+      const og = this.objectGraph;
+      let apis = this.apis_[ctorId] || [];
+      // Do not copy properties belonging to Function API from ctors.
+      const newAPIs = this.dedupFilter_(
+          apis, this.getClassAPIs_(ctorId).filter(api => ![
+            'arguments',
+            'name',
+            'length',
+            'caller',
+          ].includes(api)));
+
+      // Store all valid API names from ctorId on ctorId's data.
+      this.objectGraphApiMapper.map(ctorId, newAPIs, ctorId);
+      this.apis_[ctorId] = apis.concat(newAPIs);
+    },
+    function storeAPIsFromProto_(ctorId, protoId) {
+      const og = this.objectGraph;
+      // Gather APIs exposed in prototype chain beneath protoId.
+      let existingAPIs = this.getAllProtosAPIs_(protoId);
+      // Add to existingAPIs: APIs already registered to ctorId.
+      const ctorAPIs = this.apis_[ctorId] = this.apis_[ctorId] || [];
+      existingAPIs = this.dedupConcat_(existingAPIs, ctorAPIs);
+
+      // Add to ctorId: New APIs that:
+      // (1) Are not exposed by some interface lower level than protoId's, AND
+      // (2) Are not already on ctorId.
+      const newAPIs = this.dedupFilter_(existingAPIs,
+                                       this.getClassAPIs_(protoId));
+      this.objectGraphApiMapper.map(ctorId, newAPIs, protoId);
+      this.apis_[ctorId] = ctorAPIs.concat(newAPIs);
+    },
+    function getAllProtosAPIs_(id) {
```

Comment:
I think this produces the same list for an ID, without input from other parts of the processing. For heavily referenced prototypes, this approach will recompute the same list many times. Does it make sense to compute and store the flattened protoId to list of APIs once after all prototypes have been found?